File: Threads/C++/Threads.cpp
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <conio.h>
#include <windows.h>
#include <process.h>
#include <math.h>

int n = 100000;
int num_testes = 5;
long vetor[100000];
long resp_threadless[100000];
long resp_thread[100000];
long pos = 0;

HANDLE  hCountMutex = CreateSemaphore(NULL, 1, 1, NULL);

void buildvector() {
  srand ( time(NULL) );
  for (int i = 0; i < n; ++i) {
    vetor[i] = (rand() % n) + 1;
  }
}

DWORD WINAPI function_th( LPVOID lpParam ) {
  //printf("ID: %d\n",lpParam);
  while (true) {
    WaitForSingleObject(hCountMutex, 1);
    long j = pos;
    ++pos;
    ReleaseSemaphore(hCountMutex, 1, NULL);
     
    if (j >= n) {
      break;
    }
    
    int count = 0;
    double x = vetor[j];
    //for (int i = 0; i < n; ++i) {
    while (x > 1.0) {
      //0.97
      x = x*0.97;
      ++count;
    }
    //}
    //printf("ID: %d - %d - %d \n", lpParam, j, count);
    resp_thread[j] = count;
  }
}

void function() {
  for (long j = 0; j < n; ++j) {
    int count = 0;
    double x = vetor[j];
    while (x > 1.0) {
      x = x*0.97;
      ++count;
    }
    resp_threadless[j] = count;
  }
  
}

void threaded(int num_threads) {
  double time_start, time_stop, total_time, sum_times;
  sum_times = 0;
  printf("Threaded %d\n", num_threads);
  for (int i = 0; i < num_testes; ++i) {
    time_start = (double)clock()/CLOCKS_PER_SEC;
    
    pos = 0;
    HANDLE hThreadArray[num_threads];
    DWORD dwThreads[num_threads];
    for (int j = 0; j < num_threads; ++j) {
      hThreadArray[j] = CreateThread(NULL, 0, function_th, (void*)j, 0, &(dwThreads[j]));
    }
    WaitForMultipleObjects(num_threads, hThreadArray, TRUE, INFINITE);
    for(int j = 0; j < num_threads; ++j)
    {
      //CloseHandle(hThreadArray[j]);
    }
    
    time_stop = (double)clock()/CLOCKS_PER_SEC;
    total_time = (time_stop - time_start)*1000;
    printf("%d/%d = %f\n", i+1, num_testes, total_time);
    sum_times += total_time;
  }
  printf("Media: %f\n\n", sum_times/num_testes);
}

void threadless() {
  double time_start, time_stop, total_time, sum_times;
  sum_times = 0;
  printf("Threadless\n");
  for (int i = 0; i < num_testes; ++i) {
    time_start = (double)clock()/CLOCKS_PER_SEC;
    
    function();
    
    time_stop = (double)clock()/CLOCKS_PER_SEC;
    total_time = (time_stop - time_start)*1000;
    printf("%d/%d = %f\n", i+1, num_testes, total_time);
    sum_times += total_time;
  }
  printf("Media: %f\n\n", sum_times/num_testes);
}

int main() {
  buildvector();
  threadless();
  threaded(1);
  threaded(2);
  //threaded(3);
  threaded(4);
  //threaded(5);
  //threaded(6);
  threaded(8);
  threaded(16);
  getch();
}

File: Threads/Python/threads.py
import time
import threading
import multiprocessing
import math
import random

n = 10000
input = [0 for x in range(n)]
resp_threadless = [0 for x in range(n)]
resp_thread = [0 for x in range(n)]
hCountSemaphore = multiprocessing.Lock()
pos = 0

def buildvector():
	global input
	input = [random.randrange(0,n) for x in range(n)]

def function(x):
	count = 0
	while (x > 1.0):
		#0.97
		x = x*0.995;
		count += 1;
	return count


def test(function):
	print "Threadless"
	tests = 5
	mean = 0
	first_iteration = True
	for j in range(tests):
		begin = time.clock()
		
		for i in range(n):
			resp_threadless = function(i)
			
		end = time.clock()
		mean += (end-begin)*1000
		print "Test" + str(j) + ": " + str((end-begin)*1000)
	mean /= tests
	print "Mean: " + str(mean)
	print ""
	
def test_thread(function, num_threads):
	print "Threaded %d" % num_threads
	tests = 6
	mean = 0
	first_iteration = True

	for j in range(tests):
		pool = multiprocessing.Pool(processes=num_threads)
		begin = time.clock()
		
		pool.map(function, range(n))
		
		#global pos 
		#pos = 0
		#ths = []
		#for i in range(num_threads):
		#	th = multiprocessing.Process(target=function_th, args=(i,))
		#	ths.append(th)
		#	th.start()

		#for i in range(num_threads):
		#	th = ths[i]
		#	th.join()
		
		end = time.clock()
		if (j != 0):
		  mean += (end-begin)*1000
		print "Test" + str(j) + ": " + str((end-begin)*1000)
	mean /= (tests - 1)
	print "Mean: " + str(mean)
	print ""	

#Main

def main():
	buildvector()
	test(function)
	test_thread(function,1)
	test_thread(function,2)
	test_thread(function,4)
	test_thread(function,8)
	test_thread(function,16)
	
if __name__ == "__main__":
  main()